### PR TITLE
PathMap spec

### PR DESCRIPTION
**Pull request by chris-martin**
(In progress)

### CODE REVIEW

File: tests/Network/Wai/Middleware/OpenApi/PathMapSpec.hs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+          in  counterexample ("result = " <> show result) (result == Just True)
+
+    it "case 3" $
+      property $
+        forAll (arbitrary `suchThat` (/= "a")) $ \segment ->
+          let result =
+                PathMap.lookup
+                  (renderPath $ Path [segment])
+                  (PathMap.fromList [("/{}", False), ("/a", True)])
+          in  counterexample ("result = " <> show result) (isNothing result)
```
Comment (chris-martin) [inline tests/Network/Wai/Middleware/OpenApi/PathMapSpec.hs:49]: The funky case as described in https://github.com/freckle/wai-middleware-openapi/pull/1#discussion_r1800584356 - the spec written here describes what the code presently does, not what it should do.

File: src/Network/Wai/Middleware/OpenApi/PathMap.hs [hunk 2]
```diff
@@ -1,8 +1,10 @@
...
-  , fromOpenApi
-  , lookup
-  ) where
+module Network.Wai.Middleware.OpenApi.PathMap (
+  PathMap,
+  fromOpenApi,
+  fromList,
+  lookup,
+  empty,
+) where
```
Comment (pbrisbin) [inline src/Network/Wai/Middleware/OpenApi/PathMap.hs:7]: Whoops, thanks for adding the `fourmolu.yaml`. Can you re-run it to squash out these differences?

File: src/Network/Wai/Middleware/OpenApi/PathMap.hs [hunk 3]
```diff
@@ -17,17 +19,21 @@ import Data.OpenApi (OpenApi, PathItem)
 import Data.OpenApi qualified as OpenApi
 import System.FilePath.Posix qualified as Posix
 
-newtype PathMap = PathMap
-  { unwrap :: Map TemplatedPath PathItem
+newtype PathMap a = PathMap
```
Comment (pbrisbin) [inline src/Network/Wai/Middleware/OpenApi/PathMap.hs:22]: Is this polymorphism just to make the tests easier to write?

FWIW, I find the property-based testing hard to follow and not very good at documenting the behaviors you uncovered. I think things would be clearer if you gave yourself a helper to create concrete `PathItem`s and test against fixed scenarios with them.

File: src/Network/Wai/Middleware/OpenApi/Schema.hs [hunk 4]
```diff
@@ -44,17 +44,17 @@ data SchemaNotFound
     NonJsonResponseBody BSL.ByteString String
   deriving stock (Show)
 
-note :: MonadError e m => e -> Maybe a -> m a
+note :: (MonadError e m) => e -> Maybe a -> m a
 note e = maybe (throwError e) pure
 
-wip :: MonadError SchemaNotFound m => Maybe a -> m a
+wip :: (MonadError SchemaNotFound m) => Maybe a -> m a
 wip = maybe (throwError $ MissingPath "/todo") pure
```
Comment (pbrisbin) [inline src/Network/Wai/Middleware/OpenApi/Schema.hs]: Haha, I forgot I still had this in the main PR... Oops.

File: tests/Network/Wai/Middleware/OpenApi/PathMapSpec.hs [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+-- Failed attempt at writing a property test because I cannot find
+-- any sensible property that 'lookup' satisfies:
```
Comment (pbrisbin) [inline tests/Network/Wai/Middleware/OpenApi/PathMapSpec.hs:52]: Yeah... I don't think property tests are a good fit here. I'd rather see the behaviors as they're documented in the spec set up as concrete scenarios with test values.

File: src/Network/Wai/Middleware/OpenApi/PathMap.hs (hunk 3 above)
Comment (chris-martin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs:22]: > Is this polymorphism just to make the tests easier to write?

Yeah, it means you don't actually have to construct `PathItem`s, since that's not relevant to what's being tested, which is only how the keys match. But we could give the polymorphic type a different name and alias `type PathMap = Something PathItem` to avoid complicating the user API.

> I find the property-based testing hard to follow

Yeah, I just started down that road because I was suspecting that I needed quickcheck to generate some larger maps for me to find the problem cases. If wonky `Ord`s are a problem I don't necessarily expect to be able to find it with small examples.

Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs:22]: > If wonky `Ords` are a problem I don't necessarily expect to be able to find it with small examples

Well, you gave three cases [here](https://github.com/freckle/wai-middleware-openapi/pull/1#discussion_r1800584356). I feel like one `it` for each of those would be sufficient.

Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs:22]: > means you don't actually have to construct `PathItem`s

Well, you have to define something you can distinguish. I don't really see `mempty & summary ?~ "foo"` _that_ much more difficult than some other non-`PathItem` type (that can be distinguished). And for test that needn't distinguish (like an `isNothing` assertion), then `mempty` is about as simple as it gets.
